File: clients/rust/src/async_client.rs
//! Native async transport with cancellation by dropping the in-flight future/stream.
use crate::{
    ingestion, protocol_error, retrieval, validation, DriverError, Error, Event, IngestRequest,
    IngestResult, ProtocolInfo, Provider, Result, RetrievalDelete, RetrievalDeleteResult,
    RetrievalIndexRequest, RetrievalIndexResult, RetrievalResult, RetrievalSearch, RunRequest,
    RunResult, PROTOCOL_VERSION,
};
use reqwest::{Client as HttpClient, Response};
use serde::Deserialize;
use serde_json::Value;
use std::collections::VecDeque;
use std::time::Duration;

#[derive(Clone)]
pub struct AsyncAgenticClient {
    base: reqwest::Url,
    token: String,
    http: HttpClient,
}
impl AsyncAgenticClient {
    pub fn new(url: &str, token: impl Into<String>) -> Result<Self> {
        Self::with_ca_pem(url, token, None)
    }
    /// Add a private CA without disabling certificate or hostname verification.
    pub fn with_ca_pem(url: &str, token: impl Into<String>, ca: Option<&[u8]>) -> Result<Self> {
        let token = token.into();
        let base = crate::transport::endpoint(url, &token)?;
        let mut builder = HttpClient::builder()
            .redirect(reqwest::redirect::Policy::none())
            .retry(reqwest::retry::never())
            .connect_timeout(Duration::from_secs(10));
        if let Some(pem) = ca {
            builder = builder.add_root_certificate(reqwest::Certificate::from_pem(pem)?);
        }
        Ok(Self {
            base,
            token,
            http: builder.build()?,
        })
    }
    async fn request(&self, path: &str, body: Option<&Value>, stream: bool) -> Result<Response> {
        let url = self
            .base
            .join(path)
            .map_err(|_| Error::Protocol("Invalid endpoint."))?;
        let mut request = if let Some(data) = body {
            self.http.post(url).json(data)
        } else {
            self.http.get(url)
        };
        request = request
            .bearer_auth(&self.token)
            .header("AgenticDriver-Version", PROTOCOL_VERSION)
            .header("AgenticDriver-Accept-Optional-Events", "true")
            .header(
                "Accept",
                if stream {
                    "text/event-stream"
                } else {
                    "application/json"
                },
            );
        if body.is_none() {
            request = request.timeout(Duration::from_secs(10));
        }
        let response = request.send().await?;
        if !response.status().is_success() {
            #[derive(Deserialize)]
            struct Envelope {
                error: DriverError,
            }
            return match read_json::<Envelope>(response).await {
                Ok(payload) => Err(Error::Driver(payload.error)),
                Err(_) => Err(Error::Protocol("The driver returned an HTTP error.")),
            };
        }
        crate::transport::check_version(response.headers())?;
        Ok(response)
    }
    pub async fn ingest_context(&self, request: &IngestRequest) -> Result<IngestResult> {
        let value: Value = read_json(
            self.request(
                "v1/retrieval/ingest",
                Some(&serde_json::to_value(request)?),
                false,
            )
            .await?,
        )
        .await?;
        let (id, revision) = request.document.identity();
        if !ingestion::valid(&value["ingestion"])
            || !retrieval::receipt(&value, &request.corpus, id, revision)
        {
            return Err(protocol_error(
                "INVALID_RESPONSE",
                "Invalid ingestion provenance or a mismatched receipt.",
            ));
        }
        Ok(serde_json::from_value(value)?)
    }
    pub async fn search_context(&self, request: &RetrievalSearch) -> Result<RetrievalResult> {
        let value: Value = read_json(
            self.request(
                "v1/retrieval/search",
                Some(&serde_json::to_value(request)?),
                false,
            )
            .await?,
        )
        .await?;
        if !retrieval::valid(&value) {
            return Err(protocol_error(
                "INVALID_RESPONSE",
                "Invalid retrieval evidence.",
            ));
        }
        let result: RetrievalResult = serde_json::from_value(value)?;
        if !retrieval::selection(Some(&result), Some(request)) {
            return Err(protocol_error(
                "INVALID_RESPONSE",
                "Evidence does not match the requested scope.",
            ));
        }
        Ok(result)
    }
    pub async fn index_context(
        &self,
        request: &RetrievalIndexRequest,
    ) -> Result<RetrievalIndexResult> {
        let value: Value = read_json(
            self.request(
                "v1/retrieval/index",
                Some(&serde_json::to_value(request)?),
                false,
            )
            .await?,
        )
        .await?;
        if !retrieval::receipt(
            &value,
            &request.corpus,
            &request.source.id,
            &request.source.revision,
        ) {
            return Err(protocol_error(
                "INVALID_RESPONSE",
                "Invalid indexing receipt.",
            ));
        }
        Ok(serde_json::from_value(value)?)
    }
    pub async fn delete_context(&self, request: &RetrievalDelete) -> Result<RetrievalDeleteResult> {
        let result: RetrievalDeleteResult = read_json(
            self.request(
                "v1/retrieval/delete",
                Some(&serde_json::to_value(request)?),
                false,
            )
            .await?,
        )
        .await?;
        if &result.request != request {
            return Err(protocol_error(
                "INVALID_RESPONSE",
                "Mismatched deletion receipt.",
            ));
        }
        Ok(result)
    }
    pub async fn protocol(&self) -> Result<ProtocolInfo> {
        let info: ProtocolInfo = read_json(self.request("v1/protocol", None, false).await?).await?;
        if info.protocol != "agenticdriver"
            || info.version != PROTOCOL_VERSION
            || !info
                .supported_versions
                .iter()
                .any(|v| v == PROTOCOL_VERSION)
        {
            return Err(protocol_error(
                "INVALID_RESPONSE",
                "The driver returned an invalid protocol descriptor.",
            ));
        }
        Ok(info)
    }
    pub async fn management(&self) -> Result<crate::ManagementSnapshot> {
        let value: Value = read_json(self.request("v1/management", None, false).await?).await?;
        crate::management::snapshot(value, None)
    }
    pub async fn configure_provider(
        &self,
        input: &crate::ConfigureProvider,
    ) -> Result<crate::ManagementSnapshot> {
        let value: Value = read_json(
            self.request(
                "v1/management/providers",
                Some(&serde_json::to_value(input)?),
                false,
            )
            .await?,
        )
        .await?;
        crate::management::snapshot(value, Some(&input.provider.id))
    }
    pub async fn providers(&self) -> Result<Vec<Provider>> {
        self.provider_catalog(false).await
    }
    pub async fn refresh_providers(&self) -> Result<Vec<Provider>> {
        self.provider_catalog(true).await
    }
    async fn provider_catalog(&self, refresh: bool) -> Result<Vec<Provider>> {
        #[derive(Deserialize)]
        struct Catalog {
            providers: Vec<Provider>,
        }
        let path = if refresh {
            "v1/providers?refresh=true"
        } else {
            "v1/providers"
        };
        Ok(read_json::<Catalog>(self.request(path, None, false).await?)
            .await?
            .providers)
    }
    pub async fn submit_job(&self, request: &crate::JobSubmit) -> Result<crate::JobInfo> {
        let value: Value = read_json(
            self.request(
                "v1/jobs/submit",
                Some(&serde_json::to_value(request)?),
                false,
            )
            .await?,
        )
        .await?;
        crate::jobs::info(value, Some(request), None)
    }
    pub async fn read_job(&self, request: &crate::JobIdentity) -> Result<crate::JobInfo> {
        let value: Value = read_json(
            self.request("v1/jobs/read", Some(&serde_json::to_value(request)?), false)
                .await?,
        )
        .await?;
        crate::jobs::info(value, None, Some(request))
    }
    pub async fn cancel_job(&self, request: &crate::JobIdentity) -> Result<crate::JobInfo> {
        let value: Value = read_json(
            self.request(
                "v1/jobs/cancel",
                Some(&serde_json::to_value(request)?),
                false,
            )
            .await?,
        )
        .await?;
        crate::jobs::info(value, None, Some(request))
    }
    pub async fn job_events(
        &self,
        request: &crate::JobEventsRequest,
    ) -> Result<crate::JobEventPage> {
        let value: Value = read_json(
            self.request(
                "v1/jobs/events",
                Some(&serde_json::to_value(request)?),
                false,
            )
            .await?,
        )
        .await?;
        crate::jobs::page(value, request)
    }
    pub async fn create_session(
        &self,
        request: &crate::SessionCreate,
    ) -> Result<crate::SessionSnapshot> {
        let value: Value = read_json(
            self.request(
                "v1/sessions/create",
                Some(&serde_json::to_value(request)?),
                false,
            )
            .await?,
        )
        .await?;
        crate::sessions::snapshot(value, Some(request), None)
    }
    pub async fn read_session(
        &self,
        request: &crate::SessionIdentity,
    ) -> Result<crate::SessionSnapshot> {
        let value: Value = read_json(
            self.request(
                "v1/sessions/read",
                Some(&serde_json::to_value(request)?),
                false,
            )
            .await?,
        )
        .await?;
        crate::sessions::snapshot(value, None, Some(request))
    }
    pub async fn delete_session(
        &self,
        request: &crate::SessionIdentity,
    ) -> Result<crate::SessionDeleteResult> {
        let value: Value = read_json(
            self.request(
                "v1/sessions/delete",
                Some(&serde_json::to_value(request)?),
                false,
            )
            .await?,
        )
        .await?;
        crate::sessions::deletion(value, request)
    }
    pub async fn report_tool_progress(
        &self,
        identity: &crate::ToolExecutionIdentity,
    ) -> Result<crate::ToolExecutionReceipt> {
        let value: Value = read_json(
            self.request(
                "v1/tool-executions/progress",
                Some(&serde_json::to_value(identity)?),
                false,
            )
            .await?,
        )
        .await?;
        crate::application_tools::receipt(value, identity, crate::ToolExecutionStatus::Progress)
    }
    pub async fn complete_tool(
        &self,
        result: &crate::ToolExecutionResult,
    ) -> Result<crate::ToolExecutionReceipt> {
        let value: Value = read_json(
            self.request(
                "v1/tool-executions/results",
                Some(&serde_json::to_value(result)?),
                false,
            )
            .await?,
        )
        .await?;
        crate::application_tools::receipt(
            value,
            result.identity(),
            crate::ToolExecutionStatus::Accepted,
        )
    }
    pub async fn decide_approval(
        &self,
        decision: &crate::ApprovalDecision,
    ) -> Result<crate::ApprovalResolution> {
        let value: Value = read_json(
            self.request(
                "v1/approvals/decisions",
                Some(&serde_json::to_value(decision)?),
                false,
            )
            .await?,
        )
        .await?;
        crate::approvals::receipt(value, decision)
    }
    pub async fn run(&self, request: &RunRequest) -> Result<RunResult> {
        if !request.application_tools.is_empty() {
            return Err(protocol_error(
                "TOOL_STREAM_REQUIRED",
                "Use stream to execute application-owned tools.",
            ));
        }
        if request.approvals.is_some() {
            return Err(protocol_error(
                "APPROVAL_STREAM_REQUIRED",
                "Use stream to receive and decide interactive approvals.",
            ));
        }
        let result: RunResult = read_json(
            self.request("v1/runs", Some(&serde_json::to_value(request)?), false)
                .await?,
        )
        .await?;
        if !validation::result_valid(&result, request) {
            return Err(protocol_error(
                "INVALID_RESPONSE",
                "The driver returned an invalid run result.",
            ));
        }
        Ok(result)
    }
    /// Open a stream. Dropping it, or a pending `next()` future, closes its response.
    pub async fn stream(&self, request: &RunRequest) -> Result<EventStream> {
        let response = self
            .request("v1/runs", Some(&serde_json::to_value(request)?), true)
            .await?;
        crate::transport::check_sse(response.headers())?;
        // Retain only fields needed to verify the host's selection, not the prompt/attachments.
        let mut selection = RunRequest::new(&request.provider, &request.model, "");
        selection.retrieval.clone_from(&request.retrieval);
        selection.session.clone_from(&request.session);
        selection.approvals.clone_from(&request.approvals);
        selection
            .application_tools
            .clone_from(&request.application_tools);
        selection.tools.clone_from(&request.tools);
        Ok(EventStream {
            state: Some(StreamState {
                response,
                request: selection,
                decoder: crate::events::EventDecoder::default(),
                bytes: VecDeque::new(),
                line: Vec::new(),
                skip_lf: false,
                first_line: true,
            }),
        })
    }
}

/// An owned, backpressured event stream. No worker thread or background reader.
///
/// Failure/cancellation terminal events are delivered as `Event`s; inspect `payload()`.
/// Transport/wire failures return `Err` once, then `None`. Completion closes immediately.
/// Cancelling a polled `next()` future closes the entire stream; it cannot be resumed.
#[must_use = "A stream must be consumed to receive its result; dropping it cancels the request"]
pub struct EventStream {
    state: Option<StreamState>,
}
struct StreamState {
    response: Response,
    request: RunRequest,
    decoder: crate::events::EventDecoder,
    bytes: VecDeque<u8>,
    line: Vec<u8>,
    skip_lf: bool,
    first_line: bool,
}
impl EventStream {
    /// Close immediately; dropping the stream has the same network effect.
    pub fn close(&mut self) {
        self.state.take();
    }
    pub fn is_closed(&self) -> bool {
        self.state.is_none()
    }

    /// Receive one event. With `tokio::select!`, losing this branch closes the stream
    /// if the future was polled. Create it again only when intentionally starting a run.
    pub async fn next(&mut self) -> Option<Result<Event>> {
        // The future owns the response and all parser buffers across await points.
        // Cancellation drops them even if the caller retains the now-closed stream.
        let mut state = self.state.take()?;
        let result = state.read_event().await;
        if result.as_ref().is_ok_and(|event| !event.is_terminal()) {
            self.state = Some(state);
        }
        Some(result)
    }
}
impl StreamState {
    async fn read_event(&mut self) -> Result<Event> {
        loop {
            while let Some(byte) = self.bytes.pop_front() {
                if self.skip_lf {
                    self.skip_lf = false;
                    if byte == b'\n' {
                        continue;
                    }
                }
                if byte == b'\r' || byte == b'\n' {
                    self.skip_lf = byte == b'\r';
                    let raw = std::mem::take(&mut self.line);
                    let raw = if self.first_line {
                        self.first_line = false;
                        raw.strip_prefix(b"\xef\xbb\xbf").unwrap_or(&raw)
                    } else {
                        &raw
                    };
                    let line = std::str::from_utf8(raw).map_err(|_| {
                        protocol_error("INVALID_STREAM", "The event stream is not valid UTF-8.")
                    })?;
                    if let Some(event) = self.decoder.line(line, &self.request)? {
                        return Ok(event);
                    }
                } else {
                    self.line.push(byte);
                    if self.line.len() > validation::MAX_BYTES {
                        return Err(protocol_error(
                            "RESPONSE_TOO_LARGE",
                            "An event exceeded 2 MB.",
                        ));
                    }
                }
            }
            let Some(chunk) = self.response.chunk().await? else {
                // EOF does not dispatch a partial frame, but invalid UTF-8 still fails.
                std::str::from_utf8(&self.line).map_err(|_| {
                    protocol_error("INVALID_STREAM", "The event stream is not valid UTF-8.")
                })?;
                return Err(protocol_error(
                    "INCOMPLETE_STREAM",
                    "Connection closed before a terminal run event.",
                ));
            };
            self.bytes.extend(chunk.iter().copied());
        }
    }
}

async fn read_json<T: serde::de::DeserializeOwned>(mut response: Response) -> Result<T> {
    let mut data = Vec::new();
    while let Some(chunk) = response.chunk().await? {
        if data.len() + chunk.len() > validation::MAX_BYTES {
            return Err(protocol_error(
                "RESPONSE_TOO_LARGE",
                "Response exceeded 2 MB.",
            ));
        }
        data.extend_from_slice(&chunk);
    }
    serde_json::from_slice(&data).map_err(|_| {
        protocol_error(
            "INVALID_RESPONSE",
            "The driver returned invalid JSON or an invalid response shape.",
        )
    })
}

File: clients/python/src/agenticdriver/async_client.py
"""Native asyncio transport. Install ``agenticdriver[async]`` for HTTPX."""

import asyncio
import json
import ssl
from collections.abc import AsyncGenerator, AsyncIterator, Callable, Mapping
from contextlib import asynccontextmanager
from types import TracebackType
from typing import TYPE_CHECKING, Any, cast
from typing_extensions import Unpack

if TYPE_CHECKING:
    import httpx

from .connections import CreateInvitation, ConnectionInvitation, ConnectionCredentials, ConnectionList, connection_value
from .management import ConfigureProvider, ManagementSnapshot, snapshot as management_snapshot
from .setup import ProviderSetupRequest, ProviderSetupSnapshot, setup_request, setup_snapshot
from ._errors import DriverError
from ._protocol import (
    EventDecoder,
    SSEDecoder,
    parse_json,
    valid_catalog,
    valid_error,
    valid_protocol,
    valid_result,
    valid_resolution,
    valid_tool_receipt,
    valid_timestamp,
)
from ._sessions import valid_session_snapshot, valid_session_delete
from ._jobs import valid_job, valid_job_page
from .jobs import JobSubmit, JobIdentity, JobInfo, JobEventsRequest, JobEventPage
from .models import SessionCreate, SessionIdentity, SessionSnapshot, SessionDeleteResult
from ._retrieval import valid_retrieval, valid_index_result, valid_delete_result
from ._transport import (
    TERMINAL_EVENTS,
    check_version,
    connection_url,
    headers,
    validate_timeout,
)
from .models import ProviderInfo, ProtocolInfo, RunEvent, RunRequest, RunResult, ApprovalDecision, ApprovalResolution
from .retrieval import (
    RetrievalSearch,
    RetrievalResult,
    RetrievalIndexRequest,
    RetrievalIndexResult,
    RetrievalDelete,
    RetrievalDeleteResult,
)
from .ingestion import IngestRequest, IngestResult
from .models import ToolExecutionIdentity, ToolExecutionResult, ToolExecutionReceipt


class AsyncRunStream(AsyncIterator[RunEvent]):
    """Use ``async with`` or ``aclose()``; cancellation also closes the response."""

    def __init__(
        self,
        iterator: AsyncGenerator[RunEvent, None],
        release: Callable[["AsyncRunStream"], None],
    ):
        self._iterator = iterator
        self._release = release
        self._closed = False
        self._pending: asyncio.Future[RunEvent] | None = None

    def __aiter__(self) -> "AsyncRunStream":
        return self

    async def __anext__(self) -> RunEvent:
        if self._closed:
            raise StopAsyncIteration
        if self._pending is not None:
            raise RuntimeError("Read each stream from one task at a time.")
        pending = asyncio.ensure_future(self._iterator.__anext__())
        self._pending = pending
        try:
            event = await pending
        except BaseException:
            self._pending = None
            await self.aclose()
            raise
        finally:
            self._pending = None
        if event["type"] in TERMINAL_EVENTS:
            await self.aclose()
        return event

    async def aclose(self) -> None:
        if self._closed:
            return
        self._closed = True
        try:
            if self._pending is not None and not self._pending.done():
                self._pending.cancel()
                try:
                    await self._pending
                except asyncio.CancelledError:
                    pass
            await self._iterator.aclose()
        finally:
            self._release(self)

    async def __aenter__(self) -> "AsyncRunStream":
        if self._closed:
            raise DriverError("STREAM_CLOSED", "This stream is closed.")
        return self

    async def __aexit__(
        self,
        exc_type: type[BaseException] | None,
        exc: BaseException | None,
        tb: TracebackType | None,
    ) -> None:
        await self.aclose()


class AsyncAgenticClient:
    def __init__(
        self,
        url: str,
        token: str,
        *,
        ca_file: str | None = None,
        io_timeout: float | None = None,
        discovery_timeout: float | None = 10,
        transport: "httpx.AsyncBaseTransport | None" = None,
    ):
        self._url = connection_url(url, token)
        self._token = token
        self._io_timeout = validate_timeout(io_timeout)
        self._discovery_timeout = validate_timeout(discovery_timeout)
        try:
            import httpx
        except ImportError:
            raise ImportError(
                "Install 'agenticdriver[async]' to use AsyncAgenticClient."
            ) from None
        self._client = httpx.AsyncClient(
            verify=ssl.create_default_context(cafile=ca_file),
            timeout=None,
            follow_redirects=False,
            trust_env=False,
            transport=transport,
        )
        self._streams: set[AsyncRunStream] = set()
        self._closed = False

    def _ensure_open(self) -> None:
        if self._closed:
            raise DriverError("CLIENT_CLOSED", "This client is closed.")

    async def __aenter__(self) -> "AsyncAgenticClient":
        self._ensure_open()
        return self

    async def __aexit__(
        self,
        exc_type: type[BaseException] | None,
        exc: BaseException | None,
        tb: TracebackType | None,
    ) -> None:
        await self.aclose()

    async def aclose(self) -> None:
        self._closed = True
        try:
            for stream in tuple(self._streams):
                await stream.aclose()
        finally:
            await self._client.aclose()

    @asynccontextmanager
    async def _request(
        self, path: str, body: Mapping[str, object] | None = None, stream: bool = False
    ) -> AsyncIterator["httpx.Response"]:
        self._ensure_open()
        async with self._client.stream(
            "GET" if body is None else "POST",
            self._url + path,
            content=json.dumps(body).encode() if body is not None else None,
            headers=headers(self._token, stream, body is not None),
            timeout=self._io_timeout if body is not None else self._discovery_timeout,
        ) as response:
            if not response.is_success:
                try:
                    payload = parse_json(await self._bytes(response, 64_000)).get(
                        "error", {}
                    )
                except (DriverError, AttributeError):
                    payload = {}
                if valid_error(payload):
                    raise DriverError(
                        payload["code"],
                        payload["message"],
                        payload["retryable"],
                        payload.get("outcome"),
                    )
                raise DriverError(
                    "HTTP_ERROR",
                    f"Driver returned HTTP {response.status_code}.",
                    response.status_code == 429 or response.status_code >= 500,
                )
            check_version(response.headers.get_list("AgenticDriver-Version") or None)
            yield response

    @staticmethod
    async def _bytes(response: "httpx.Response", limit: int = 2_000_000) -> bytes:
        data = bytearray()
        async for chunk in response.aiter_bytes():
            if len(data) + len(chunk) > limit:
                raise DriverError(
                    "RESPONSE_TOO_LARGE", "The response exceeded its size limit."
                )
            data.extend(chunk)
        return bytes(data)

    async def _json(self, response: "httpx.Response") -> Any:
        return parse_json(await self._bytes(response))

    async def report_tool_progress(self, message: ToolExecutionIdentity) -> ToolExecutionReceipt:
        async with self._request("v1/tool-executions/progress", message) as response:
            result = await self._json(response)
            if not valid_tool_receipt(result, message, "progress"):
                raise DriverError("INVALID_RESPONSE", "The tool receipt does not match its submission; reconcile the originating run.")
            return cast(ToolExecutionReceipt, result)

    async def complete_tool(self, message: ToolExecutionResult) -> ToolExecutionReceipt:
        async with self._request("v1/tool-executions/results", message) as response:
            result = await self._json(response)
            if not valid_tool_receipt(result, message, "accepted"):
                raise DriverError("INVALID_RESPONSE", "The tool receipt does not match its submission; reconcile the originating run.")
            return cast(ToolExecutionReceipt, result)

    async def decide_approval(self, decision: ApprovalDecision) -> ApprovalResolution:
        async with self._request("v1/approvals/decisions", decision) as response:
            result = await self._json(response)
            if not valid_resolution(result, decision=decision):
                raise DriverError("INVALID_RESPONSE", "The approval receipt does not match the decision; reconcile using the run stream.")
            return cast(ApprovalResolution, result)

    async def ingest_context(self, request: IngestRequest) -> IngestResult:
        async with self._request("v1/retrieval/ingest", request) as response:
            result = await self._json(response)
            document = request["document"]
            source = document if document["type"] == "reference" else document["source"]
            if (
                not valid_index_result(
                    result, {"corpus": request["corpus"], "source": source}
                )
                or "ingestion" not in result
            ):
                raise DriverError(
                    "INVALID_RESPONSE",
                    "The driver returned invalid ingestion provenance or a mismatched receipt.",
                )
            return cast(IngestResult, result)

    async def search_context(self, request: RetrievalSearch) -> RetrievalResult:
        async with self._request("v1/retrieval/search", request) as response:
            result = await self._json(response)
            if not valid_retrieval(result, request):
                raise DriverError(
                    "INVALID_RESPONSE",
                    "The driver returned invalid or unscoped retrieval evidence.",
                )
            return cast(RetrievalResult, result)

    async def index_context(
        self, request: RetrievalIndexRequest
    ) -> RetrievalIndexResult:
        async with self._request("v1/retrieval/index", request) as response:
            result = await self._json(response)
            if not valid_index_result(result, request):
                raise DriverError(
                    "INVALID_RESPONSE",
                    "The driver returned an invalid indexing receipt.",
                )
            return cast(RetrievalIndexResult, result)

    async def delete_context(self, request: RetrievalDelete) -> RetrievalDeleteResult:
        async with self._request("v1/retrieval/delete", request) as response:
            result = await self._json(response)
            if not valid_delete_result(result, request):
                raise DriverError(
                    "INVALID_RESPONSE",
                    "The driver returned an invalid deletion receipt.",
                )
            return cast(RetrievalDeleteResult, result)

    async def submit_job(self, request: JobSubmit) -> JobInfo:
        return await self._job_request("submit", request)

    async def read_job(self, request: JobIdentity) -> JobInfo:
        return await self._job_request("read", request)

    async def cancel_job(self, request: JobIdentity) -> JobInfo:
        return await self._job_request("cancel", request)

    async def _job_request(self, operation: str, request: JobSubmit | JobIdentity) -> JobInfo:
        async with self._request(f"v1/jobs/{operation}", request) as response:
            result = await self._json(response)
            if not valid_job(result, request):
                raise DriverError("INVALID_RESPONSE", "Invalid or mismatched job metadata.")
            return cast(JobInfo, result)

    async def job_events(self, request: JobEventsRequest) -> JobEventPage:
        async with self._request("v1/jobs/events", request) as response:
            result = await self._json(response)
            if not valid_job_page(result, request):
                raise DriverError("INVALID_RESPONSE", "Invalid or out-of-order job event page.")
            return cast(JobEventPage, result)

    async def create_session(self, request: SessionCreate) -> SessionSnapshot:
        async with self._request("v1/sessions/create", request) as response:
            result = await self._json(response)
            if not valid_session_snapshot(result, request, "create", valid_timestamp):
                raise DriverError("INVALID_RESPONSE", "The conversation response does not match its request.")
            return cast(SessionSnapshot, result)

    async def read_session(self, request: SessionIdentity) -> SessionSnapshot:
        async with self._request("v1/sessions/read", request) as response:
            result = await self._json(response)
            if not valid_session_snapshot(result, request, "read", valid_timestamp):
                raise DriverError("INVALID_RESPONSE", "The conversation response does not match its request.")
            return cast(SessionSnapshot, result)

    async def delete_session(self, request: SessionIdentity) -> SessionDeleteResult:
        async with self._request("v1/sessions/delete", request) as response:
            result = await self._json(response)
            if not valid_session_delete(result, request):
                raise DriverError("INVALID_RESPONSE", "The conversation response does not match its request.")
            return cast(SessionDeleteResult, result)

    async def create_invitation(self, request: CreateInvitation) -> ConnectionInvitation:
        async with self._request("v1/management/invitations", request) as response:
            return cast(ConnectionInvitation, connection_value(await self._json(response), "code"))

    async def exchange_connection(self) -> ConnectionCredentials:
        async with self._request("v1/connections/exchange", {}) as response:
            return cast(ConnectionCredentials, connection_value(await self._json(response), "token"))

    async def connections(self) -> ConnectionList:
        async with self._request("v1/management/connections") as response:
            return cast(ConnectionList, connection_value(await self._json(response), "list"))

    async def revoke_connection(self, connection_id: str) -> bool:
        async with self._request("v1/management/connections/revoke", {"id": connection_id}) as response:
            return bool(connection_value(await self._json(response), "revoke")["revoked"])

    async def provider_setup(self, request: ProviderSetupRequest) -> ProviderSetupSnapshot:
        async with self._request("v1/management/setup", setup_request(request)) as response:
            return setup_snapshot(await self._json(response), request)

    async def management(self) -> ManagementSnapshot:
        async with self._request("v1/management") as response:
            return management_snapshot(await self._json(response))

    async def configure_provider(self, request: ConfigureProvider) -> ManagementSnapshot:
        async with self._request("v1/management/providers", request) as response:
            return management_snapshot(await self._json(response), request["provider"]["id"], request.get("remove") is True)

    async def providers(self, *, refresh: bool = False) -> list[ProviderInfo]:
        async with self._request(
            "v1/providers?refresh=true" if refresh else "v1/providers"
        ) as response:
            catalog = await self._json(response)
            if not valid_catalog(catalog):
                raise DriverError(
                    "INVALID_RESPONSE",
                    "The driver returned an invalid provider catalog.",
                )
            return cast(list[ProviderInfo], catalog["providers"])

    async def protocol(self) -> ProtocolInfo:
        async with self._request("v1/protocol") as response:
            info = await self._json(response)
            if not valid_protocol(info):
                raise DriverError(
                    "INVALID_RESPONSE",
                    "The driver returned an invalid protocol descriptor.",
                )
            return cast(ProtocolInfo, info)

    async def run(self, **request: Unpack[RunRequest]) -> RunResult:
        if request.get("applicationTools"):
            raise DriverError("TOOL_STREAM_REQUIRED", "Use stream() to execute application-owned tools.")
        if request.get("approvals"):
            raise DriverError("APPROVAL_STREAM_REQUIRED", "Use stream() to receive and decide interactive approvals.")
        async with self._request("v1/runs", request) as response:
            result = await self._json(response)
            if not valid_result(result, request):
                raise DriverError(
                    "INVALID_RESPONSE", "The driver returned an invalid run result."
                )
            return cast(RunResult, result)

    def stream(self, **request: Unpack[RunRequest]) -> AsyncRunStream:
        self._ensure_open()
        stream = AsyncRunStream(self._events(request), self._streams.discard)
        self._streams.add(stream)
        return stream

    async def _events(self, request: RunRequest) -> AsyncGenerator[RunEvent, None]:
        async with self._request("v1/runs", request, stream=True) as response:
            if "text/event-stream" not in response.headers.get("Content-Type", ""):
                raise DriverError("INVALID_RESPONSE", "Expected an SSE response.")
            decoder, frames = EventDecoder(request), SSEDecoder()
            async for chunk in response.aiter_bytes():
                for data in frames.feed(chunk):
                    event = decoder.accept(parse_json(data, "INVALID_STREAM"))
                    if event is not None:
                        yield cast(RunEvent, event)
                        if event["type"] in TERMINAL_EVENTS:
                            return
            frames.finish()
